Add image code Refactor

dakaneye opened the pull request:
This is an attempt to modularize and unit test the add-image workflow (by tag/digest)
* This should _not_ be modifying the add-from-archive or import-image-content workflow
* This should not be changing _ANY_ functionality

dakaneye commented on anchore_engine/services/catalog/catalog_impl.py:
```diff
@@ -1826,6 +1873,30 @@ def perform_policy_evaluation(
     return curr_evaluation_record, curr_evaluation_result
 
 
+ImageKey = collections.namedtuple("ImageKey", ["tag", "digest"])
```
this bit is new, and meant to support other features who _have_ an image tag and digest, and want it to be easy to add an image for analysis

dakaneye replied:
Note: this flow will NOT create any subscriptions for the image, that bit happens currently in the external API service

Vijay-P replied:
any reason why we don't use dataclasses instead of namedtuples? we're using 3.8 now so compatibility isn't an issue anymore.

Vijay-P replied:
https://medium.com/@jacktator/dataclass-vs-namedtuple-vs-object-for-performance-optimization-in-python-691e234253b9

dakaneye replied:
I didn't even know about data classes! I used this just because I was more familiar. There are some python 3.6 relics left in engine (around integration tests, I believe), so I am reticent to adding this now, but super glad you shared this so I know about it for the future!

alfredodeza commented on anchore_engine/services/catalog/catalog_impl.py:
```diff
@@ -483,201 +482,240 @@ def image(dbsession, request_inputs, bodycontent=None):
+            err.message, in_httpcode=err.__response_code__, details=err.detail
+        )
+        httpcode = err.__response_code__
+    except Exception as err:
```
I think we should not allow blanket exceptions anymore as part of new additions https://github.com/anchore/anchore-engine/issues/867

dakaneye replied:
I think specifically, I want to leave this error because there are lots of possibilities of errors to be thrown here, but this should be a target for a later refactor

alfredodeza replied:
Unfortunately that doesn't end up working well for testing, debugging or how the rest of the code-base deals with exceptions. The `image()` function gets called in a couple of places. One is in `list_images()` and the other one in `add_image()`, *both of which* do enormous try/except blocks that are also catching anything that might get raised. 
As I explain in #867, this broad exception catching at multiple levels in the stack makes it impossible to debug, and causes problems for testing.

I'm going to be a -1 on adding more broad-exception catches like this one. There needs to be a solid reason for catching everything, while avoiding multiple broad try/except blocks that are catching exceptions.

dakaneye replied:
@alfredodeza I totally agree with you here, but I have two comments:
* I'm not adding this exception block (it's not new), just ported over as I modularized things
* If I change this logic now...I need to go through each sub method, identify _all possible_ errors, and be 100% sure that expected use cases don't end up missing a more specific block. At this point, that's not a risk I'm looking to take as this is a highly used path in the engine. At this point I am just trying to do the bare bones modularization...but you are 100% right that this is bad and needs to change, I'm just trying to avoid doing that change in this PR.

alfredodeza commented on anchore_engine/services/catalog/catalog_impl.py:
```diff
@@ -483,201 +482,240 @@ def image(dbsession, request_inputs, bodycontent=None):
-                                    "tag": image_info["fulltag"],
-                                },
-                            )
+    except AnchoreApiError:
+        raise
+    except Exception as err:
+        logger.exception("Error adding image")
+        raise err
```
The only thing that this block is adding is a log entry mentioning that there is an error adding an image. The cost is high: indents the code into a large try/except block, makes it hard to debug, and it not necessary. Similar as previous comment, for reference, issue: https://github.com/anchore/anchore-engine/issues/867

alfredodeza commented on anchore_engine/services/catalog/catalog_impl.py:
```diff
@@ -483,201 +482,240 @@ def image(dbsession, request_inputs, bodycontent=None):
+        )
+        try:
+            add_event(fail_event, dbsession)
+        except Exception:
+            logger.warn("Ignoring error creating image registry lookup event")
+        raise err
```
similar to other broad exception comments

dakaneye replied:
This one looks like it's squelching an error when a "fail" event _fails_ to be submitted....and might need some thinking about re-implementation, because removing this WOULD change the functionality here (main goal was to refactor and modularize)

alfredodeza commented on anchore_engine/services/catalog/catalog_impl.py:
```diff
@@ -483,201 +482,240 @@ def image(dbsession, request_inputs, bodycontent=None):
+            manifest = json.dumps(image_info["manifest"])
+        else:
+            raise Exception("no manifest from get_image_info")
+    except Exception as err:
+        raise Exception("could not fetch/parse manifest - exception: " + str(err))
```
This is similar to the other exception handling, but causes a problem: the original exception is being caught and only added in its string representation to a new error that adds a message, obfuscating the originating exception. This will cause debugging to be really hard

alfredodeza commented:
I guess I didn't read the subject of the PR that well that says that nothing should be changing functionality :( Feel free to ignore the exception comments if they don't apply @dakaneye

dakaneye commented:
> I guess I didn't read the subject of the PR that well that says that nothing should be changing functionality :( Feel free to ignore the exception comments if they don't apply @dakaneye

I think we could address the exception handling possibly, @zhill what do you think?

Vijay-P commented on anchore_engine/services/catalog/catalog_impl.py:
```diff
@@ -483,201 +482,240 @@ def image(dbsession, request_inputs, bodycontent=None):
 
         elif method == "POST":
-            timer = time.time()
+            return_object, httpcode = image_post(
```
if we're going this far, might as well split the handlers for get and post into two different methods, create a common pre-processing method, and update the calls in anchore_engine/services/catalog/api/controllers/default_controller.py. will be easier to read. having the POST handler tacked on to the end of the images() method feels like an afterthought

dakaneye replied:
What would you think about refactoring this in a subsequent change? Wanted to minimize the impact here as there _is_ some pre-processing that is common for the two methods for now

Vijay-P replied:
That's fine w me.

zhill commented:
> > I guess I didn't read the subject of the PR that well that says that nothing should be changing functionality :( Feel free to ignore the exception comments if they don't apply @dakaneye
> 
> I think we could address the exception handling possibly, @zhill what do you think? I was trying to be risk averse here but it seems like a worthwhile improvement.

I think it is good to address the cases in the direct code paths changed here, but we should be careful of scope-creep if we start delving into the methods called. Agreed they all need review, but should be done in small focused updates if we start modifying functions called here but not modified in this PR.

dakaneye commented:
@alfredodeza @Vijay-P @zhill I addressed some of the comments and updated the branch against master. LMK what you think. 

Some of the exception blocks are hard to remove (not impossible) but I _did_ want to limit the scope of the change here as much as possible. So I did what I could for now, and happy to open more follow up tech debt tasks.

Vijay-P reviewed (approved):
LGTM, but would like to see more work done to this part of the code in the near future.
